src/Dapr.Client/DaprClientGrpc.cs

```diff
@@ -294,10 +358,123 @@ public override async ValueTask<IReadOnlyList<BulkStateItem>> GetBulkStateAsync(
                 Message = invokeRequest,
             };
 
-            return await this.MakeGrpcCallHandleError(
-                 options => client.InvokeServiceAsync(request, options),
-                 headers,
-                 cancellationToken);
+            var callOptions = new CallOptions(headers: headers ?? new Metadata(), cancellationToken: cancellationToken);
+
+            // add token for dapr api token based authentication
+            var daprApiToken = Environment.GetEnvironmentVariable("DAPR_API_TOKEN");
+
+            if (daprApiToken != null)
+            {
+                callOptions.Headers.Add("dapr-api-token", daprApiToken);
+            }
+
+            return (request, callOptions);
+        }
+
+        private async Task<InvocationResponse<TResponse>> MakeInvokeRequestAsyncWithResponse<TRequest, TResponse>(
+            InvocationRequest<TRequest> request,
+            CancellationToken cancellationToken = default)
+        {
+            
+            
+            Any serializedData = null;
+            if (request.Body != null)
+            {
+                if (typeof(TResponse) == typeof(byte[]))
+                {
+                    // User has passed in raw bytes
+                    var requestBytes = (byte[])Convert.ChangeType(request.Body, typeof(byte[]));
+                    serializedData = new Any { Value = ByteString.CopyFrom(requestBytes), TypeUrl = typeof(byte[]).FullName };
+                }
+                else
+                {
+                    serializedData = TypeConverters.ToAny(request.Body, this.jsonSerializerOptions);
+                }
+            }
+
+            try
+            {
+                var invokeResponse = new InvocationResponse<TResponse>();
+                Autogenerated.InvokeServiceRequest invokeRequest;
+                CallOptions callOptions;
+                (invokeRequest, callOptions) = this.MakeInvokeRequestAsync(request.AppId, request.MethodName, serializedData, request.HttpExtension, cancellationToken);
+                var grpcCall = client.InvokeServiceAsync(invokeRequest, callOptions);
+
+                var response = await grpcCall.ResponseAsync;
+                var responseHeaders = await grpcCall.ResponseHeadersAsync;
+                var trailers = grpcCall.GetTrailers();
+                var grpcStatus = grpcCall.GetStatus();
+
+                var headers = grpcCall.ResponseHeadersAsync.Result.ToDictionary(kv => kv.Key, kv => kv.ValueBytes);
+
+                if (typeof(TResponse) == typeof(byte[]))
+                {
+                    // User wants to receive raw bytes
+                    var responseBytes = new byte[response.Data.Value.Length];
+                    response.Data.Value?.CopyTo(responseBytes, 0);
+                    invokeResponse.Body = (TResponse)(response.Data.Value.IsEmpty ? default : Convert.ChangeType(responseBytes, typeof(TResponse)));
```

Comment:
You can remove `Convert.ChangeType` here and just do a cast. You already know that `TResponse` is the same type as `byte[]`. `Convert.ChangeType` is for converting values between representations - like converting a floating point number to a decimal.